kanidmd/src/lib/credential/mod.rs

```diff
@@ -706,6 +768,59 @@ impl Credential {
         }
     }
 
+    pub(crate) fn update_backup_code(&self, backup_code: HashSet<String>) -> Self {
+        let type_ = match &self.type_ {
+            // TODO: we should panic/reject CredentialType != PasswordMfa?
+            CredentialType::Password(pw) | CredentialType::GeneratedPassword(pw) => {
+                debug_assert!(false);
+                CredentialType::PasswordMfa(
+                    pw.clone(),
+                    None,
+                    Map::new(),
+                    Some(BackupCode {
+                        code_set: backup_code,
+                    }),
+                )
+            }
+            CredentialType::PasswordMfa(pw, totp, wan, _) => CredentialType::PasswordMfa(
+                pw.clone(),
+                totp.clone(),
+                wan.clone(),
+                Some(BackupCode {
+                    code_set: backup_code,
+                }),
+            ),
+            CredentialType::Webauthn(wan) => {
+                debug_assert!(false);
+                CredentialType::Webauthn(wan.clone())
+            }
+        };
+        Credential {
+            type_,
+            claims: self.claims.clone(),
+            uuid: self.uuid,
+        }
+    }
+
+    // pub(crate) fn remove_backup_code(&self) -> Self {
```

Comment:
I think you should uncomment this, it's good :) 

If you make it so only PasswordMfa takes backup code, it's a much simpler code path. You could probabry consider returning Err() when not PMFA here though so that we respond that the transform is invalid though.